addons/account/models/account_reconcile_model.py

```diff
@@ -812,19 +812,26 @@ def _get_invoice_matching_amls_candidates(self, st_line, partner):
                 }
 
         # Search without any matching based on textual information.
-        if partner:
+        if self.matching_order == 'new_first':
+            order = 'date_maturity DESC, date DESC, id DESC'
+        else:
+            order = 'date_maturity ASC, date ASC, id ASC'
 
-            if self.matching_order == 'new_first':
-                order = 'date_maturity DESC, date DESC, id DESC'
+        if not partner:
+            st_line_currency = st_line.foreign_currency_id or st_line.journal_id.currency_id or st_line.company_currency_id
+            if self.company_id.currency_id == st_line_currency:
+                aml_amount_field = 'balance'
             else:
-                order = 'date_maturity ASC, date ASC, id ASC'
-
-            amls = self.env['account.move.line'].search(aml_domain, order=order)
-            if amls:
-                return {
-                    'allow_auto_reconcile': False,
-                    'amls': amls,
-                }
+                aml_amount_field = 'amount_currency'
+            aml_domain += [
+                (aml_amount_field, '=', -st_line.amount_residual),
+            ]
```

Comment:
```suggestion
if not partner:
    st_line_currency = st_line.foreign_currency_id or st_line.journal_id.currency_id or st_line.company_currency_id
    if st_line_currency == self.company_id.currency_id:
        aml_domain = [
            ('currency_id', '=', st_line_currency.id),
            ('amount_residual', '=', -st_line.amount_residual),
        ]
    else:
        aml_domain = [
            ('currency_id', '=', st_line_currency.id),
            ('amount_residual_currency', '=', -st_line.amount_residual),
        ]
```

Reply:
Ok for the amount, but the currency is already set by the `match_currency` settings
https://github.com/odoo/odoo/blob/41ca2fcf0449e9b30075808734984b032cb70901/addons/account/models/account_reconcile_model.py#L658-L659

Reply:
@agr-odoo Only when `match_same_currency` is ticked

Reply:
If it's not ticked, why do we want to match the currency here?